crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs

```diff
@@ -0,0 +1,3321 @@
+            CallbackNode::ArrowFunction(func) => Some(&func.body),
+        }
+    }
+}
+
+impl<'a> GetSpan for CallbackNode<'a> {
+    fn span(&self) -> Span {
+        match self {
+            CallbackNode::Function(func) => func.span,
+            CallbackNode::ArrowFunction(func) => func.span,
+        }
+    }
+}
+
+// https://github.com/facebook/react/blob/1b0132c05acabae5aebd32c2cadddfb16bda70bc/packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js#L1789
+fn get_reactive_hook_callback_index(node: &CallExpression) -> Option<usize> {
+    let node_name = get_node_name_without_react_namespace(&node.callee);
+
+    let hook_name = node_name?;
+
+    match hook_name.as_str() {
+        "useEffect" | "useLayoutEffect" | "useCallback" | "useMemo" => Some(0),
+        "useImperativeHandle" => Some(1),
+        _ => {
+            // TODO: custom nodes
+            None
+        }
+    }
+}
+
+fn get_node_name_without_react_namespace<'a, 'b>(expr: &'b Expression<'a>) -> Option<&'b Atom<'a>> {
+    match expr {
+        Expression::StaticMemberExpression(member) => {
+            if let Expression::Identifier(_ident) = &member.object {
+                return Some(&member.property.name);
+            }
+            None
+        }
+        Expression::Identifier(ident) => Some(&ident.name),
+        _ => None,
+    }
+}
+
+#[derive(Debug)]
+struct Dependency<'a> {
+    iref: &'a IdentifierReference<'a>,
+    // the symbol id that this dependency is referring to
+    symbol_id: Option<SymbolId>,
+    chain: Vec<String>,
+}
+
+impl Hash for Dependency<'_> {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.iref.name.hash(state);
+        self.symbol_id.hash(state);
+        self.chain.hash(state);
+    }
+}
+
+impl PartialEq for Dependency<'_> {
+    fn eq(&self, other: &Self) -> bool {
+        self.iref.name == other.iref.name && self.chain == other.chain
+    }
+}
+
+impl Eq for Dependency<'_> {}
+
+impl Dependency<'_> {
+    fn to_string(&self) -> String {
+        [vec![self.iref.name.to_string()], self.chain.clone()].concat().join(".")
+    }
+
+    fn contains(&self, other: &Self) -> bool {
+        self.iref.name == other.iref.name && chain_contains(&self.chain, &other.chain)
+    }
+}
+
+fn chain_contains(a: &[String], b: &[String]) -> bool {
+    for (index, part) in b.iter().enumerate() {
+        let Some(other) = a.get(index) else { return false };
+        if other != part {
+            return false;
+        };
+    }
+
+    true
+}
+
+fn analyze_property_chain<'a, 'b>(
+    expr: &'b Expression<'a>,
+    ctx: &'b LintContext<'a>,
+) -> Result<Option<Dependency<'a>>, ()> {
+    match expr {
+        Expression::Identifier(ident) => {
+            // TODO: @overlookmotel how can i avoid the unsafe???
+            let ident = unsafe {
+                std::mem::transmute::<&IdentifierReference<'_>, &'a IdentifierReference<'a>>(ident)
+            };
```

Comment:
any ideas on how to get around this?

Reply:
@overlookmotel / @Boshen  ? 🙂

Reply:
Store the name / span / id directly inside `Dependency`?

Reply:
good shout 🙂